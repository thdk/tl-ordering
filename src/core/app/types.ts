import { Action } from "redux";
import { IOrderItem, IOrderItemState } from "../orderitems/types";
import { IOrder, IOrderState } from "../orders/types";
import { IProductState } from "../products/types";

export interface IState {
    orders: IOrderState;
    orderItems: IOrderItemState;
    products: IProductState;
}

export interface IAction<T, P = undefined> extends Action<T> {
    payload?: P;
}

export interface IOrderWithData extends IOrder {
    items: IOrderItemWithPrice[];
    total: number;
}

export interface IOrderItemWithPrice extends IOrderItem {
    readonly unitPrice: number;
}
