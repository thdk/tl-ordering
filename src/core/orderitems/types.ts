export interface IOrderItemDictionary { [orderId: string]: IOrderItem[]; }

export interface IOrderItemState { byOrderId: IOrderItemDictionary; }

export interface IOrderItem {
    readonly productId: string;
    readonly quantity: number;
}

// Types for deserialized api data

export interface IOrderItemData {
    productId: string;
    quantity: number;
    orderId: string;
}

// Types representing data from/for API

interface IApiData {
    [manyProps: string]: any;
}

// "product-id": "B102",
// "quantity": "5",
// "unit-price": "4.99",
// "total": "24.95"
export interface IApiOrderItem extends IApiData {
    quantity: string;
    total: string;
}
