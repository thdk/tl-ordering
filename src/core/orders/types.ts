import { IApiOrderItem, IOrderItemData } from "../orderitems/types";

export interface IOrderState {
    byId: IOrderDictionary;
    visibleIds: string[];
    isLoading: boolean;
}

export interface IOrderDictionary { [orderId: string]: IOrder; }

export interface IOrder {
    readonly id: string;
    readonly customerId: string;
    readonly isPlaced?: boolean;
}

// Types for deserialized api data

export interface IOrderData {
    id: string;
    items: IOrderItemData[];
    customerId: string;
}

export interface IPlaceOrderData {
    success: boolean;
    order: IOrderData;
    reason?: string;
}

// Types representing data from/for API

interface IApiData {
    [manyProps: string]: any;
}

// {
//  "id": "2",
//   "customer-id": "2",
//   "items": [
//     {
//       "product-id": "B102",
//       "quantity": "5",
//       "unit-price": "4.99",
//       "total": "24.95"
//     }
//   ],
//   "total": "24.95"
// }
export interface IApiOrder extends IApiData {
    id: string;
    items: IApiOrderItem[];
    total: string;
}

export interface IApiPlaceOrderResult extends IApiData {
    result: string;
    order: IApiOrder;
    reason?: string;
}
