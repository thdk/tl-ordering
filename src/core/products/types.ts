export interface IProductDictionary { [productId: string]: IProduct; }

export interface IProductState {
    byId: IProductDictionary;
    allIds: string[];
    isLoading: boolean;
}

export enum ProductCategory {
    "unknown" = 0,
    "tools" = 1,
    "electronics" = 2,
}

export interface IProduct {
    readonly id: string;
    readonly description: string;
    readonly category: ProductCategory;
    readonly price: number;
}

// Types for deserialized api data

export interface IProductData {
    readonly id: string;
    readonly description: string;
    readonly category: ProductCategory;
    readonly price: number;
}

// Types representing data from/for API

export interface IApiProduct {
    readonly id: string;
    readonly description: string;
    readonly category: string;
    readonly price: string;
}
