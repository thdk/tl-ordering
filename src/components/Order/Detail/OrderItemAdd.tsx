import React, { useRef } from "react";
import { connect } from "react-redux";
import { Dispatch } from "redux";

import { IState } from "../../../core/app/types";
import { addOrderItem } from "../../../core/orderitems/actions";
import { IOrderItem } from "../../../core/orderitems/types";
import { IProduct } from "../../../core/products/types";

export interface IPropsFromState {
    products: IProduct[];
}

export interface IPropsFromDispatch {
    readonly onAdd: (item: IOrderItem) => void;
}

export interface IOrderItemAddProps {
    readonly orderId: string;
}

type Props = IPropsFromState & IPropsFromDispatch & IOrderItemAddProps;

const OrderItemAdd = (props: Props) => {
    const { onAdd, products } = props;
    const defaultQuantiy = 1;

    const productIdRef = useRef<HTMLSelectElement | null>(null);
    const quantityRef = useRef<HTMLInputElement | null>(null);

    const handleSubmit = (e: React.FormEvent) => {
        e.preventDefault();
        e.stopPropagation();

        // Todo: validate input
        if (productIdRef.current && quantityRef.current) {
            if (productIdRef.current.value) {
                onAdd({
                    productId: productIdRef.current.value,
                    quantity: +quantityRef.current.value,
                });

                productIdRef.current.value = "";
                quantityRef.current.value = defaultQuantiy.toString();
            }

        }
    };
    return (
        <form onSubmit={handleSubmit}>
            <label>Product id:</label><select ref={productIdRef}>
                ${products.map((product) => <option key={product.id} value={product.id}>{product.id}</option>)}
            </select>
            <label>Quantity:</label><input type="number" min="0" ref={quantityRef} defaultValue={"1"}></input>
            <input type="submit" value="Add product" />
        </form>
    );
};

const mapStateToProps = (state: IState) => {
    return {
        products: state.products.allIds.map((id) => state.products.byId[id]),
    };
};

const mapDispatchToProps = (dispatch: Dispatch, ownProps: IOrderItemAddProps) => {
    return {
        onAdd: (item: IOrderItem) => {
            dispatch(addOrderItem(item, ownProps.orderId));
        },
    };
};

const ConnectedOrderItemAdd = connect(
    mapStateToProps,
    mapDispatchToProps,
)(OrderItemAdd);

export default ConnectedOrderItemAdd;
