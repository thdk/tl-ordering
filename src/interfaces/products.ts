export interface IProduct {
    id: string;
    unitPrice: number;
}