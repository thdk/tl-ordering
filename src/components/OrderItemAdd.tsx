import React, { useRef } from 'react'
import { addOrderItem } from '../actions/orderItems';
import { connect } from 'react-redux';
import { IOrderItem } from '../interfaces/orders';
import { Dispatch } from 'redux';
import { IState } from '../interfaces/state';
import { IProduct } from '../interfaces/products';

export type PropsFromState = {
    products: IProduct[];
};

export type PropsFromDispatch = {
    readonly onAdd: (item: IOrderItem) => void;
}

export interface IOrderItemAddProps {
    readonly orderId: string;
}

type Props = PropsFromState & PropsFromDispatch & IOrderItemAddProps;

const OrderItemAdd = (props: Props) => {
    const { onAdd, products } = props;
    const defaultQuantiy = 1;

    const productIdRef = useRef<HTMLSelectElement | null>(null);
    const quantityRef = useRef<HTMLInputElement | null>(null);

    const handleSubmit = (e: React.FormEvent) => {
        e.preventDefault();
        e.stopPropagation();

        // Todo: validate input
        if (productIdRef.current && quantityRef.current) {
            onAdd({
                productId: productIdRef.current.value,
                quantity: +quantityRef.current.value,
            });

            productIdRef.current.value = "";
            quantityRef.current.value = defaultQuantiy.toString();
        }
    };
    return (
        <form onSubmit={handleSubmit}>
            <label>Product id:</label><select ref={productIdRef}>
                ${products.map(product => <option key={product.id} value={product.id}>{product.id}</option>)}
            </select>
            <label>Quantity:</label><input type="number" min="0" ref={quantityRef} defaultValue={"1"}></input>
            <input type="submit" value="Add product" />
        </form>
    );
};

const mapStateToProps = (state: IState) => {
    return {
        products: state.products.allIds.map(id => state.products.byId[id])
    }
}

const mapDispatchToProps = (dispatch: Dispatch, ownProps: IOrderItemAddProps) => {
    return {
        onAdd: (item: IOrderItem) => {
            dispatch(addOrderItem(item, ownProps.orderId))
        }
    }
};

const ConnectedOrderItemAdd = connect(
    mapStateToProps,
    mapDispatchToProps,
)(OrderItemAdd);

export default ConnectedOrderItemAdd;