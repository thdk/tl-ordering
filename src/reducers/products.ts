import { IProductState } from "../interfaces/state";

export function products(state: IProductState = { byId: {}, allIds: [] }) {
    return state;
}