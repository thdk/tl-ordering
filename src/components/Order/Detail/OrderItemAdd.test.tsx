import { shallow } from "enzyme";
import React from "react";

import { IState } from "../../../core/app/types";
import { ProductCategory } from "../../../core/products/types";
import { mapStateToProps, OrderItemAdd } from "./OrderItemAdd";

describe("Rendering OrderItemAdd", () => {
    const onAdd = jest.fn();

    const component = shallow(
        <OrderItemAdd
            orderId={"o1"}
            products={[
                {
                    id: "p1",
                    category: ProductCategory.electronics,
                    description: "Product description 1",
                    price: 5,
                },
                {
                    id: "p2",
                    category: ProductCategory.electronics,
                    description: "Product description 2",
                    price: 5,
                },
                {
                    id: "p3",
                    category: ProductCategory.electronics,
                    description: "Product description 3",
                    price: 5,
                },
            ]}
            onAdd={onAdd}
        />,
    );

    it("should match snapshot", () => {
        expect(component.debug()).toMatchSnapshot();
    });

    it("should call mockFunction on button click", () => {
        // Try to add product without selecting product from dropdown first
        component.find("form").simulate("submit", { preventDefault: jest.fn(), stopPropagation: jest.fn() });
        expect(onAdd).not.toBeCalledWith();

        // Now select a product from the dropdown list
        component.find("select").simulate("change", { currentTarget: { value: "p1" } });
        component.find(".orderitem-add-quantity").simulate("change", { currentTarget: { value: "3" } });

        component.find("form").simulate("submit", { preventDefault: jest.fn(), stopPropagation: jest.fn() });

        expect(onAdd).toBeCalledWith("o1", { productId: "p1", quantity: 3 });
    });
});

describe("OrderItemAdd mapStateToProps", () => {
    const stateProps = mapStateToProps({
        orderItems: {
            byOrderId: {},
        },
        orders: {
            byId: {},
            isLoading: false,
            visibleIds: [],
        },
        products: {
            allIds: ["p1"],
            byId: {
                p1: {
                    category: ProductCategory.electronics,
                    description: "Product one",
                    id: "p1",
                    price: 10,
                },
            },
            isLoading: false,
        },
    });

    expect(stateProps.products.length).toBe(1);
});
