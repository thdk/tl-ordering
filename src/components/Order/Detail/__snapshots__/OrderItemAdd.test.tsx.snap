// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering OrderItemAdd should match snapshot 1`] = `
"<form onSubmit={[Function]}>
  <label>
    Product id:
  </label>
  <select value=\\"\\" onChange={[Function]} className=\\"orderitem-add-product-select\\">
    <option value=\\"\\">
      Select a product
    </option>
    $
    <option value=\\"p1\\">
      p1
    </option>
    <option value=\\"p2\\">
      p2
    </option>
    <option value=\\"p3\\">
      p3
    </option>
  </select>
  <label>
    Quantity:
  </label>
  <input className=\\"orderitem-add-quantity\\" onChange={[Function]} type=\\"number\\" min=\\"0\\" defaultValue=\\"1\\" />
  <input className=\\"orderitem-add-submit-button\\" type=\\"submit\\" value=\\"Add product\\" />
</form>"
`;
